text/0264-registry-schema-codegen.md

```diff
@@ -0,0 +1,303 @@
+  } ]
+},
+```
+
+To indicate that *exactly one* of the properties must be specified (note how `properties`
+has been factored out and applies to both branches of the `oneOf`).
+
+> This is a weird way to specify it but I imagine it works because if you
+> specify both properties you could satisfy both subschemas which is not allowed
+> by `oneOf`.
+
+There are also some technical problems with the spec that need to be addressed
+(or reckoned with) before we can properly consume the spec:
+
+* The CloudFormation Registry Schema is a confusing superset of JSON Schema,
+  as it combines both *properties* and *attributes* in the same type definition.
+  There are fields alongside the type definition which tell you which properties
+  are actually input properties, and which ones represent output attributes (
+  retrievable via `{ Ref }` and `{ Fn::GetAtt }`). Code generators and compatibility
+  checkers need to clearly separate these sets of properties out because they
+  need to be treated very differently.
+* There is currently no good source of information in the CloudFormation Registry
+  Schema on attributes that can be `{ Fn::GetAtt }`ed. `readOnlyProperties`
+  would seem to represent those, but it does not currently handle the case
+  where a `Property` and `Attribute` have the same name. Support for that
+  will be forthcoming, but until then we can't reliably switch over.
+
+## SAM Specification
+
+We are also generating L1s for the SAM resource types, which are by now
+being used in 60% of all customer stacks.
+
+The SAM specification has been hand-written to mimic the old CloudFormation
+specification, but with extensions. This specification is not maintained
+by CloudFormation and hence is not being converted to JSON schema by them.
+
+We will need to write our own CloudFormation specification -> JSON Schema
+converter to be able to keep on using SAM resources once we switch over.
+
+## Major Components
+
+Implementing support for this requires the following components:
+
+* Centralized spec repository with a way to patch the spec. There *will* be
+  breaking changes coming in as updates are released and upstream developers
+  make mistakes, and we still need a way of patching around those to unblock
+  our build. Since a number of resources are automatically being converted
+  from old format to new format, they are missing additional information. We
+  can enhance the spec with information obtained from the [cfn-lint] project.
+* A tool to two versions of the spec. This is necessary to generate the
+  changelog every time we import a new version of the resources. It should be
+  able to identify breaking changes, so that we can extend this tool to
+  resource developers and they can prevent themselves from accidentally
+  releasing breaking changes (reducing churn on downstream systems).
+* A tool to generate code the type and resource definitions found
+  in the schema. This can mostly be based on `json2jsii`, although we will
```

Comment:
Let's definitely try to bake as much of the capabilities into Json2jsii so we get the most leverage for other CDKs (and jsii-based projects).